examples/src/main/java/com/squareup/moshi/recipes/EnvelopeJsonAdapterFactory.java

```diff
@@ -0,0 +1,101 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.recipes;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonQualifier;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+final class EnvelopeJsonAdapterFactory implements JsonAdapter.Factory {
+  private static final class Train {
+    String name;
+  }
+
+  public static void main(String[] args) {
+    Enveloped annotation = new Enveloped() {
+      @Override public Class<? extends Annotation> annotationType() {
+        return Enveloped.class;
+      }
+    };
+    Moshi moshi = new Moshi.Builder().add(new EnvelopeJsonAdapterFactory()).build();
+    JsonAdapter<Train> adapter = moshi.adapter(Train.class, Collections.singleton(annotation));
+    Train out;
+    try {
+      out = adapter.fromJson("{\"data\":{\"name\":\"James\"}}");
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    }
+    System.out.println(out.name);
+  }
+
+  @Retention(RUNTIME) @JsonQualifier @interface Enveloped {
+  }
+
+  private static final class Envelope<T> {
+    final T data;
+
+    Envelope(T data) {
+      this.data = data;
+    }
+  }
+
+  @Override
+  public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
+    if (!hasJsonQualifier(annotations, Enveloped.class)) {
+      return null;
+    }
+    Type envelope =
+        Types.newParameterizedTypeWithOwner(EnvelopeJsonAdapterFactory.class, Envelope.class,
+            Types.getRawType(type));
+    final JsonAdapter<Envelope<?>> delegate =
+        moshi.nextAdapter(this, envelope, Collections.<Annotation>emptySet());
+    return new JsonAdapter<Object>() {
+      @Override public Object fromJson(JsonReader reader) throws IOException {
+        return delegate.fromJson(reader).data;
+      }
+
+      @Override public void toJson(JsonWriter writer, Object value) throws IOException {
+        delegate.toJson(new Envelope<>(value));
+      }
+    };
+  }
+
+  private static boolean hasJsonQualifier(Set<? extends Annotation> annotations,
+      Class<? extends Annotation> jsonQualifier) {
+    if (!jsonQualifier.isAnnotationPresent(JsonQualifier.class)) {
```

Comment:
Callers should be able to assume that Moshi is doing this check for them already.